osgp/platform/osgp-secret-management/src/main/java/org/opensmartgridplatform/secretmanagement/application/endpoints/SecretManagementEndpoint.java

```diff
@@ -37,86 +45,147 @@
-        List<TypedSecret> typedSecrets = secretManagementService.retrieveSecrets(request.getDeviceId(),
+        final List<SecretType> secretTypeList = this.converter.convertToSecretTypes(soapSecretTypes);
+        final List<TypedSecret> typedSecrets = this.secretManagementService.retrieveSecrets(request.getDeviceId(),
                 secretTypeList);
-
-        TypedSecrets soapTypedSecrets = converter.convertToSoapTypedSecrets(typedSecrets);
-
+        final TypedSecrets soapTypedSecrets = this.converter.convertToSoapTypedSecrets(typedSecrets);
         response.setTypedSecrets(soapTypedSecrets);
         response.setResult(OsgpResultType.OK);
-
         log.trace(response.toString());
+        return response;
+    }
 
+    @PayloadRoot(namespace = NAMESPACE_URI, localPart = "getNewSecretsRequest")
+    @ResponsePayload
+    public GetNewSecretsResponse getNewSecretsRequest(@RequestPayload final GetNewSecretsRequest request) throws OsgpException {
+        log.info("Handling incoming SOAP request 'getSecretsRequest' for device {}", request.getDeviceId());
+        if (log.isDebugEnabled()) {
+            log.debug(this.requestToString(request));
+        }
+        final GetNewSecretsResponse response = new GetNewSecretsResponse();
+        final SecretTypes soapSecretTypes = request.getSecretTypes();
+        if (soapSecretTypes == null) {
+            throw new TechnicalException("Missing input: secret types");
+        }
+        final List<SecretType> secretTypeList = this.converter.convertToSecretTypes(soapSecretTypes);
+        final List<TypedSecret> typedSecrets = this.secretManagementService.retrieveNewSecrets(request.getDeviceId(),
+                secretTypeList);
+        final TypedSecrets soapTypedSecrets = this.converter.convertToSoapTypedSecrets(typedSecrets);
+        response.setTypedSecrets(soapTypedSecrets);
+        response.setResult(OsgpResultType.OK);
+        log.trace(response.toString());
         return response;
     }
 
     @PayloadRoot(namespace = NAMESPACE_URI, localPart = "storeSecretsRequest")
     @ResponsePayload
-    public StoreSecretsResponse storeSecretsRequest(@RequestPayload StoreSecretsRequest request) throws OsgpException {
-
+    public StoreSecretsResponse storeSecretsRequest(@RequestPayload final StoreSecretsRequest request)
+            throws OsgpException {
         log.info("Handling incoming SOAP request 'storeSecretsRequest' for device {}", request.getDeviceId());
         log.trace(request.toString());
-
-        StoreSecretsResponse response = new StoreSecretsResponse();
-
-        TypedSecrets soapTypedSecrets = request.getTypedSecrets();
-
+        final StoreSecretsResponse response = new StoreSecretsResponse();
+        final TypedSecrets soapTypedSecrets = request.getTypedSecrets();
         if (soapTypedSecrets == null) {
             throw new TechnicalException("Missing input: typed secrets");
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Define a constant instead of duplicating this literal "Missing input: typed secrets" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3AS1192)